### PR TITLE
Hydrate iOS chats only when selected

### PR DESCRIPTION
**Pull request by sachaservan**
## Summary
- publish cloud, local, and project sidebar rows from encrypted chat indexes without decrypting full histories at startup
- materialize only selected, streaming, recovery, blank, or explicitly mutated chats
- add cancellable generation-safe selection with explicit loading/error UI and disabled message actions until hydration completes
- persist paginated rows before publishing summaries and preserve loaded pages across sync refreshes
- make title/project/delete mutations transactional against real stored chats

## Safety invariants
- no partial or summary-shaped Chat values are ever saved
- stale A→B loads and delete-vs-hydration completions cannot install the wrong chat
- pagination never advances past failed download persistence
- title/project backup happens only after a throwing local save succeeds
- current/streaming/recovery chats remain materialized; inactive chats are evicted
- anonymous chat re-encryption remains retryable through sign-in, periodic, foreground, and full sync

## Testing
- adds index-only startup, selection fence, storage routing, reconciliation, pagination persistence, mutation gate, delete failure, project transition, and anonymous retry policy tests
- `git diff --check`
- multiple static blocker/regression review passes completed
- builds and tests were not run from the CLI per `AGENTS.md`; please validate in Xcode

## Stack
- depends on #264

---
## Summary by cubic
Hydrates iOS chats only when selected to cut startup time and memory. Previously all chats decrypted at launch; now the sidebar renders from encrypted indexes and chats hydrate on selection, with actions disabled until hydration completes and a loading/error overlay with retry.

- Publishes paginated summary rows from encrypted indexes; persists pages before publishing; preserves loaded pages across sync; stops pagination on persistence failure; surfaces cancelled/failed page flags.
- Materializes only selected, streaming, recovery, or explicitly mutated chats; maintains ordered streaming snapshots; evicts inactive ones; adds a generation‑safe selection fence and per‑chat mutation gate to block stale installs and concurrent mutations.
- Cloud sync, deletion, and search: honors authoritative deletions and prevents resurrection; keeps deleted IDs until account state is cleared; posts `cloudChatRemoteDeleteCompleted` (keys: `chatId`, `userId`) on remote delete; rolls back tombstoned remote writes; remote search persistence is tombstone‑aware and prefers local on freshness conflicts; clears project state during account teardown.
- Bulk delete: suspends uploads via `quiesceUploadsForBulkDelete(userId:)`; guards `deleteAllFromCloud(userId:)` by the active account and matching `userId`; resumes with `resumeUploadsAfterBulkDelete(userId:)`; resets suspension on account teardown; makes `UploadCoalescer.clear()` async and drains tasks. Required migration: call `quiesceUploadsForBulkDelete(userId:)` → `deleteAllFromCloud(userId:)` → `resumeUploadsAfterBulkDelete(userId:)` in order; await `uploadCoalescer.clear()`.
- UI: sidebar and project pages render `ChatListSummary` rows; search can open a summary‑only chat; overlays show loading/errors with Try Again; message input and toolbar buttons disable until hydration completes; selection stays stable during chat moves.
- Tests add `LazyChatHydrationTests` covering pagination persistence/cancellation, authoritative index reconciliation, tombstone rollbacks, remote search persistence, delete‑all coordination, lazy hydration edge cases, and error ordering.

<sup>Written for commit 4e2374f112d5f6f73e3d242215ae93d10a3c91cd. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/tinfoilsh/tinfoil-ios/pull/268?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: TinfoilChat/Views/ChatSidebar.swift [hunk 1]
```diff
@@ -411,8 +419,10 @@ struct ChatSidebar: View {
                     },
                     onEdit: {
                         if editingChatId == chat.id {
-                            viewModel.updateChatTitle(chat.id, newTitle: editingTitle)
-                            editingChatId = nil
+                            Task {
+                                await viewModel.updateChatTitle(chat.id, newTitle: editingTitle)
+                                editingChatId = nil
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Views/ChatSidebar.swift]: Fixed in 35d844e: title completion clears the editor only when the same chat is still being edited.

File: TinfoilChat/ViewModels/ChatViewModel.swift [hunk 2]
```diff
@@ -1339,11 +1385,52 @@ class ChatViewModel: ObservableObject {
...
-        selectChat(chat)
+    func openProjectChat(_ chat: ChatListSummary) {
+        _ = selectChat(id: chat.id, isLocalOnly: false)
         isViewingProjectChat = true
     }
 
+    func openSummaryChat(id: String, projectId: String?, isLocalOnly: Bool) {
+        guard let projectId, activeProject?.id != projectId else {
+            if projectId == nil {
+                beginSelection(id: id)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/ViewModels/ChatViewModel.swift:1767]: Not changing this: the three selection paths have materially different project/search/storage work. A generic async selection callback would hide ownership rules without fixing a demonstrated issue.

File: TinfoilChat/ViewModels/ChatViewModel.swift [hunk 3]
```diff
@@ -5153,119 +5651,38 @@ class ChatViewModel: ObservableObject {
...
-            // Only first page loaded - include first page AND current chat if it exists
-            var syncedChatsToShow = Array(syncedChats.prefix(Constants.Pagination.chatsPerPage))
-            if let currentChatId = currentChatId,
-               let currentChat = syncedChats.first(where: { $0.id == currentChatId }),
-               !syncedChatsToShow.contains(where: { $0.id == currentChatId }) {
-                syncedChatsToShow.append(currentChat)
+        do {
+            let totalEntries = try await refreshCloudSummaryIndex(userId: userId)
+            evictInactiveMaterializedChats()
+            if totalEntries > cloudSidebarSummaries.filter({ !$0.isBlankChat }).count {
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/ViewModels/ChatViewModel.swift]: Fixed in 35d844e: Load More compares the root total against only nonblank, nonproject root summaries.

File: TinfoilChat/Services/ChatLoadingService.swift [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    static func persist(
+        _ movedChat: Chat,
+        wasLocal: Bool,
+        userId: String,
+        loadingService: any ChatLoadingService,
+        validateAccount: () throws -> Void
+    ) async throws {
+        try await loadingService.saveChat(movedChat, userId: userId, storage: .cloud)
+        do {
+            try validateAccount()
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/ChatLoadingService.swift:408]: Fixed in 35d844e: project storage validation no longer throws after both persistence mutations have committed, so ViewModel state cannot be left behind successful storage.

File: TinfoilChat/Services/CloudSyncService.swift [hunk 5]
```diff
@@ -1238,14 +1238,14 @@ class CloudSyncService: ObservableObject {
                     continuationToken: continuationToken
                 )
                 guard generation == accountGeneration else {
-                    return PaginatedChatsResult(chats: [], hasMore: false, nextToken: nil)
+                    return PaginatedChatsResult(chats: [], hasMore: false, nextToken: nil, failed: true)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/CloudSyncService.swift]: Fixed in 35d844e: pagination now distinguishes cancellation from request failure, so store wipes leave page state untouched without a false retry error.

File: TinfoilChat/Services/ChatLoadingService.swift [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+        } catch {
+            if wasLocal {
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/ChatLoadingService.swift:410]: Not changing this: the rollback branches represent different commit stages and conditions. Keeping them explicit makes the transaction easier to audit than a generalized helper.

File: TinfoilChat/Services/CloudSyncService.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/CloudSyncService.swift]: Not changing this: the repeated value is a small explicit result construction. A helper would be style-only and does not improve pagination correctness.

File: TinfoilChat/Views/ChatListView.swift [hunk 7]
```diff
@@ -235,6 +235,22 @@ struct ChatListView: View {
             viewModel.isScrollInteractionActive = false
             scrollToUserTrigger = UUID()
         }
+        .overlay {
+            if viewModel.hydratingChatId != nil || viewModel.chatHydrationError != nil {
+                VStack(spacing: 12) {
+                    if viewModel.hydratingChatId != nil {
+                        ProgressView()
+                    }
+                    Text(viewModel.chatHydrationError ?? "Loading conversation...")
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Views/ChatListView.swift:252]: Fixed in 35d844e: hydration errors retain the failed identity and show an explicit Try Again action; retries remain user initiated.

File: TinfoilChat/ViewModels/ChatViewModel.swift [hunk 8]
```diff
@@ -4017,7 +4399,12 @@ class ChatViewModel: ObservableObject {
                 await self.cloudSync.backupChat(latestChat.id)
             }
 
-            if let syncedChat = await Chat.loadChat(chatId: latestChat.id, userId: self.currentUserId) {
+            if let userId = self.currentUserId,
+               let syncedChat = try? await self.chatLoadingService.loadChat(
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/ViewModels/ChatViewModel.swift]: Fixed in 35d844e: post-stream reload now reports real storage errors and ignores expected cancellation.

File: TinfoilChatTests/LazyChatHydrationTests.swift [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+
+        #expect(await service.savedChat(id: original.id, storage: .cloud)?.title == original.title)
+        #expect(await service.counts().deletes == 0)
+    }
+
+    @Test
+    func deletingUnmaterializedIdentityDoesNotLoadChat() async throws {
+        let service = RecordingChatLoadingService()
+
+        try await service.deleteChat(id: "delete", userId: "user", storage: .cloud)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChatTests/LazyChatHydrationTests.swift]: Fixed in 35d844e: removed the tautological mock-only deletion test.

File: TinfoilChat/ViewModels/ChatViewModel.swift [hunk 10]
```diff
@@ -1266,61 +1310,63 @@ class ChatViewModel: ObservableObject {
...
         projectError = nil
         var loaded = false
+        defer {
+            if generation == projectLoadGeneration {
+                isLoadingProject = false
+            }
+        }
         do {
             let project = try await projectStorage.getProject(projectId)
+            guard !Task.isCancelled, generation == projectLoadGeneration else { return false }
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/ViewModels/ChatViewModel.swift]: Not changing this: these guards are localized and document the project generation boundary. Removing synchronous checks is cleanup without a practical behavior or performance impact.

File: TinfoilChat/Services/ChatLoadingService.swift [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+    @MainActor
+    static func saveCloudChats(
+        _ chats: [Chat],
+        userId: String,
+        loadingService: any ChatLoadingService
+    ) async -> (saved: [Chat], failedIds: [String]) {
+        var saved: [Chat] = []
+        var failedIds: [String] = []
+        for chat in chats {
+            guard !Task.isCancelled else { break }
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/ChatLoadingService.swift:159]: Not changing this: cancellation checks around each awaited save prevent old-account persisted rows from being published. The caller guard alone cannot protect per-row account transitions.

File: TinfoilChat/Services/ChatLoadingService.swift [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+        userId: String,
+        loadingService: any ChatLoadingService,
+        validateAccount: () throws -> Void
+    ) async throws {
+        try await loadingService.saveChat(movedChat, userId: userId, storage: .cloud)
+        do {
+            try validateAccount()
+        } catch {
+            if wasLocal {
+                try? await loadingService.deleteChat(id: movedChat.id, userId: userId, storage: .cloud)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/ChatLoadingService.swift]: Fixed in 6f003d4: rollback failures now preserve both the primary transition error and compensating cleanup error so the operation remains visibly retryable.

File: TinfoilChat/ViewModels/ChatViewModel.swift [hunk 13]
```diff
@@ -1695,13 +1824,107 @@ class ChatViewModel: ObservableObject {
...
+        pendingSearchResultChatId = nil
+        selectedChatId = id
+        hydratingChatId = id
+        chatHydrationError = nil
+        currentChat = nil
+        evictInactiveMaterializedChats()
+    }
+
+    func retryFailedChatHydration() {
+        guard let failedChatHydration, chatHydrationError != nil else { return }
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/ViewModels/ChatViewModel.swift]: Fixed in 6f003d4: retryFailedChatHydration now uses the computed eligibility property as its single guard.

File: TinfoilChat/ViewModels/ChatViewModel.swift [hunk 14]
```diff
@@ -4043,6 +4457,10 @@ class ChatViewModel: ObservableObject {
                 if self.currentChat?.id == latestChat.id {
                     self.currentChat = syncedChat
                 }
+            } catch is CancellationError {
+            } catch {
+                guard self.currentUserId == userId else { return }
+                self.syncErrors.append(error.localizedDescription)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/ViewModels/ChatViewModel.swift:5009]: Fixed in 6f003d4: post-stream reload errors are ignored when deletion/mutation is active or the chat no longer exists in that storage.

File: TinfoilChat/Services/CloudSyncService.swift [hunk 15]
```diff
@@ -1267,14 +1267,14 @@ class CloudSyncService: ObservableObject {
                     continuationToken: continuationToken
                 )
                 guard generation == accountGeneration else {
-                    return PaginatedChatsResult(chats: [], hasMore: false, nextToken: nil)
+                    return PaginatedChatsResult(chats: [], hasMore: false, nextToken: nil, cancelled: true)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/CloudSyncService.swift:1319]: Not changing this: the explicit cancelled/failed constructors are small and readable; extracting factories is style-only and does not improve behavior.

File: TinfoilChat/Services/ChatLoadingService.swift [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+    let hasMore: Bool
+}
+
+enum ChatPaginationCoordinator {
+    static func state(
+        original: ChatPaginationPageState,
+        next: ChatPaginationPageState,
+        allRowsPersisted: Bool,
+        pageFailed: Bool = false,
+        pageCancelled: Bool = false
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/ChatLoadingService.swift:353]: Fixed in 6f003d4: the production pagination coordinator now receives result.cancelled, keeping its cancellation policy wired to the call site.

File: TinfoilChat/Services/ChatLoadingService.swift [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+        }
+        guard wasLocal else { return }
+
+        do {
+            try await loadingService.deleteChat(id: movedChat.id, userId: userId, storage: .local)
+        } catch {
+            let localDeleteError = error
+            try await loadingService.deleteChat(id: movedChat.id, userId: userId, storage: .cloud)
+            throw localDeleteError
+        }
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/ChatLoadingService.swift:431]: Fixed in 6f003d4 with a composite rollback error: the original local-delete failure remains available without silently discarding a failed compensating cleanup.
